krm-functions/lib/nad/v1/nad.go

```diff
@@ -0,0 +1,370 @@
+		if nadConfigStruct.Plugins == nil || len(nadConfigStruct.Plugins) == 0 {
+			nadConfigStruct.Plugins = []PluginCniType{
+				{
+					Capabilities: Capabilities{
+						Ips: true,
+					},
+					Mode: ModeBridge,
+					Ipam: Ipam{
+						Type: StaticNadType,
+						Addresses: []Addresses{
+							{},
+						},
+					},
+				},
+			}
+		}
+		return nadConfigStruct, nil
+	}
+	return nadConfigStruct, nil
+}
+
+func (r *NadStruct) setNadConfig(config NadConfig) error {
+	b, err := json.Marshal(config)
+	if err != nil {
+		return err
+	}
+	return r.K.SetNestedString(string(b), ConfigType...)
+}
+
+func (r *NadStruct) getStringValue(fields ...string) string {
+	s, ok, err := r.K.NestedString(fields...)
+	if err != nil {
+		return ""
+	}
+	if !ok {
+		return ""
+	}
+	return s
+}
+
+func (r *NadStruct) GetConfigSpec() string {
+	return r.getStringValue(ConfigType...)
+}
+
+func (r *NadStruct) GetCNIType() (string, error) {
+	existingNadConfig, err := r.getNadConfig()
+	if err != nil {
+		return "", err
+	}
+	for _, plugin := range existingNadConfig.Plugins {
+		if plugin.Type == TuningType {
+			continue
+		} else {
+			return plugin.Type, nil
+		}
+	}
+	return "", nil
+}
+
+func (r *NadStruct) GetVlan() (int, error) {
+	existingNadConfig, err := r.getNadConfig()
+	if err != nil {
+		return 0, err
+	}
+	return existingNadConfig.Vlan, nil
+}
+
+func (r *NadStruct) GetNadMaster() (string, error) {
+	existingNadConfig, err := r.getNadConfig()
+	if err != nil {
+		return "", err
+	}
+	for _, plugin := range existingNadConfig.Plugins {
+		if plugin.Type == TuningType {
+			continue
+		} else {
+			return plugin.Master, nil
+		}
+	}
+	return "", nil
+}
+
+func (r *NadStruct) GetIpamAddress() ([]Addresses, error) {
+	existingNadConfig, err := r.getNadConfig()
+	if err != nil {
+		return []Addresses{}, err
+	}
+	for _, plugin := range existingNadConfig.Plugins {
+		if plugin.Type == TuningType {
+			continue
+		} else {
+			return plugin.Ipam.Addresses, nil
+		}
+	}
+	return []Addresses{}, nil
+}
+
+// SetConfigSpec sets the spec attributes in the kubeobject according the go struct
+func (r *NadStruct) SetConfigSpec(spec *nadv1.NetworkAttachmentDefinitionSpec) error {
+	return r.K.SetNestedString(spec.Config, ConfigType...)
+}
+
+func (r *NadStruct) SetCNIType(cniType string) error {
+	if cniType == "" {
+		return fmt.Errorf("unknown cniType")
+	} else {
+		if cniType == "ipvlan" {
+			r.CniSpecType = IpVlanType
+		} else if cniType == "macvlan" {
+			r.CniSpecType = MacVlanType
+		} else if cniType == "sriov" {
+			r.CniSpecType = SriovType
+		}
+		nadConfigStruct, err := r.getNadConfig()
+		if err != nil {
+			return err
+		}
+		for i, plugin := range nadConfigStruct.Plugins {
+			if plugin.Type == TuningType {
+				continue
+			} else {
+				nadConfigStruct.Plugins[i].Type = cniType
+			}
+		}
+		return r.setNadConfig(nadConfigStruct)
+	}
+}
+
+func (r *NadStruct) SetVlan(vlanType int) error {
+	if vlanType != 0 {
+		nadConfigStruct, err := r.getNadConfig()
```

Comment:
better to return err and fail fast

Reply:
changes here: https://github.com/nokia/nephio-nephio/pull/27/files